src/qutip_qip/algorithms/qpe.py

```diff
@@ -0,0 +1,212 @@
+import numpy as np
+from qutip import Qobj
+from qutip_qip.operations import Gate, ControlledGate
+from qutip_qip.circuit import QubitCircuit
+from qutip_qip.decompose import decompose_one_qubit_gate
+
+__all__ = ["qpe", "CustomGate"]
+
+
+class CustomGate(Gate):
+    """
+    Custom gate that wraps an arbitrary quantum operator.
+    """
+    def __init__(self, targets, U, **kwargs):
+        super().__init__(targets=targets, **kwargs)
+        self.targets = targets if isinstance(targets, list) else [targets]
+        self.U = U  
+        self.kwargs = kwargs
+        self.latex_str = r"U"  
+        
+    def get_compact_qobj(self):
+        return self.U
+
+
+def create_controlled_unitary(controls, targets, U, control_value=1):
+    """
+    Create a controlled unitary gate.
+    
+    Parameters
+    ----------
+    controls : list
+        Control qubits
+    targets : list
+        Target qubits
+    U : Qobj
+        Unitary operator to apply on target qubits
+    control_value : int, optional
+        Control value (default: 1)
+        
+    Returns
+    -------
+    ControlledGate
+        The controlled unitary gate
+    """
+    return ControlledGate(
+        controls=controls,
+        targets=targets,
+        control_value=control_value,
+        target_gate=CustomGate,
+        U=U
+    )
+
+
+def _cphase_to_cnot(targets, controls, arg_value):
+    """
+    Decompose a controlled phase gate into CNOTs and single-qubit gates.
+    
+    Parameters
+    ----------
+    targets : list
+        Target qubits
+    controls : list
+        Control qubits
+    arg_value : float
+        Phase angle
+        
+    Returns
+    -------
+    list
+        List of decomposed gates
+    """
+    rotation = Qobj([[1.0, 0.0], [0.0, np.exp(1.0j * arg_value)]])
+    decomposed_gates = list(
+        decompose_one_qubit_gate(rotation, method="ZYZ_PauliX")
+    )
+    new_gates = []
+    
+    gate = decomposed_gates[0]
+    gate.targets = targets
+    new_gates.append(gate)
+    
+    new_gates.append(Gate("CNOT", targets=targets, controls=controls))
+    
+    gate = decomposed_gates[4]
+    gate.targets = targets
+    new_gates.append(gate)
+    
+    new_gates.append(Gate("CNOT", targets=targets, controls=controls))
+    
+    new_gates.append(Gate("RZ", targets=controls, arg_value=arg_value / 2))
+    
+    gate = decomposed_gates[7]
+    gate.arg_value += arg_value / 4
+    new_gates.append(gate)
+    
+    return new_gates
+
+
+def inverse_qft_gate_sequence(N=1, swapping=True, to_cnot=False):
```

Comment:
This looks nice, do you think maybe we should move it to `qft.py`? Add an additional keyword argument `inverse` to `qft_gate_sequence`?

Reply:
I transferred the function to the `qft` file. Let me now if that seems okay?